### ISSUE TITLE
DSC terminology

### ISSUE THREAD

**Issue by gaow** [stephenslab/dsc2#68]
This page has to be re-defined (and everywhere in documentation and code ...):

https://stephenslab.github.io/dsc-wiki/doc/documentation/Terminology.html

**Comment by jdblischak** [comment]
Since we are trying to agree on terminology, it may be helpful to have another meeting where we try to agree on some terms. It's difficult to discuss terminology without a defined terminlogy :-)

**Comment by gaow** [comment]
@stephens999 comments here: 

https://github.com/stephenslab/dsc-wiki/blob/master/development/initial_thoughts_on_terminology_and_extraction.md

My first thoughts after reading through it is that we might have to think whether or not YAML based syntax has to be changed. I think most, if not all, features required in this document is already available under the current framework, but with perhaps counter-intuitive terminology and complicated syntax. On the other hand I do not think making the syntax entirely R-like is good because even if I write something that mimics and R parser, it is not R interpreter -- but users might think otherwise and therefore try / demand unreasonable stuff. 

Currently DSC operates in 3 steps: execute, apply tags and extract results. Or 2 steps if tags are added via automatic annotation. I started from YAML-like syntax because I think its simplicity is beyond compare and I tried to code DSC logic under this syntax. Generally I do not feel very much constrained though it does look a bit ugly in complicated cases -- mostly because it cannot be a one liner! If we are talking about re-designing syntax we should make sure the re-design is not only easier to specify DSC but also easier to extract from DSC. Or a balance between these.

The output data-frame part particularly needs more careful design. A way to go is perhaps write that part truly in R that takes input from DSC extracted data. The extracted data is basically a list of output from desired modules, of whatever return type. From here on it should be easy to entirely rely on R to operate on it. It's also more modular.

**Comment by stephens999** [comment]
thanks @gaow .
I don't think YAML syntax will need to be changed much. But I wanted to first start by 
making sure we are agreed on the main concepts, and a common terminology, and what
we would want to be able to extract. Then we can return to the YAML syntax.

My vision and hope
is that we should be able to develop a 2 step procedure (1. run pipelines; 2. extract results for plotting etc) that will work in most cases. If necessary we could add 
an additional annotation step between these two steps for edge cases (but I'm not yet convinced
it will be necessary).

I also believe that using YAML for the first step and R for the second step is an attractive initial goal.
It could be useful down the line to develop other interfaces for the second step (eg python etc).
But one thing at a time!

For now i want to check 
i) do you have suggested changes to terminology?
ii) are there any errors or misunderstandings in my text? 
iii) do you have other types of data extraction you think we should be considering that I did not mention here?

**Comment by gaow** [comment]
Thank you @stephens999 

> do you have suggested changes to terminology?

Instead of proposing chance let me try to merge your suggestion here with @jdblischak suggestion in #69 . From what you've described the concept of `module` sounds like a replacement to `step` in my old terms. My old `step` can take many parameter combinations, ie, many modules in your terms. The upside is its simplicity to easily try out various parameter combinations. The downside is the need of annotations later if only one set of combinations is required. In fact the same logic holds for the use of `block`, a simple way to organize many steps and compose pipeline. But if we do give up the succinctness of `step` we might well also give up the same property with `block` so that we remove the need of complicated annotations. Then we can replace `block` with the term `class` that @jdblischak proposes. For example the old syntax:

```
block_name:
   exec: SFA.R
   seed: 1, 2
   params:
     K: 10, 50
     tmp_file: File(sfa)
```

The new  (YAML-like) syntax:

```
class_name:
   seed: 1, 2 
   SFA_1:
     exec: SFA.R
     K: 10
     tmp_file: File(sfa)
   SFA_2(SFA_1):
     K: 20
  return: ...
```
And the pipeline will be defined by chains of logic operations on classes.

The `super class` concept that @jdblischak suggested is, to me, just an alias or short cut that helps with `DSC::run` syntax that we can certainly easily implement:

```
super_class: class_1 * (class_2, class_3)
```

Here `module` restricts choice of parameters to only one set and there is no long the concept of `step`. Because the same `module` can have many seeds then `task` would refer to each replicate of a `module`. 

From the engineering point of view, one parameter-set per module makes DSC much less efficient ... but that's a problem that I should face.

> are there any errors or misunderstandings in my text?

Conceptually no as I take the syntax you suggested are only conceptual. But you said "with exception of "multi-modules" (see later)." which I did not find elsewhere. multi-modules is essentially the old step design, ie, multiple parameter combinations per exec, right? I find it is hard to keep both multi-module and module concepts because multi-module will surely require annotation tags to track it.

> do you have other types of data extraction you think we should be considering that I did not mention here?

Dataframe sounds quite reasonable. Also if we can make dataframes then it makes it very easy to use Python on it as there is a package for Python interface to R dataframe (Hadley wrote the R part; my interface is more general yet less efficient). I'm not very much worried about formatting extracted data for now, though, because whatever final output we need will be extracted from a well-defined R list in a single RDS file as a result of `dsc -e`, so it'd be a pure R data manipulation problem which I believe R experts in the lab can give great suggestions. The harder part is how to extract data using as little (or no) annotation as possible. This goes back to question i) the terminology and syntax redesign. 

To sum up, to my understanding we use `superclass/class/module` model to replace `block/step` model. Certainly `module` will retire the use of annotation. But apparently where there are many exec and parameter combinations the new syntax lacks succinctness. For example the old syntax below would yield very long text in the new design proposed in this comment:

```
block_name:
  exec: 1.R, 2.R, 3.R, 4.R, 5.R
  params:
     K: 1, 2, 4, 50, 100, 200, 300
```

**Comment by stephens999** [comment]
sorry, @gaow  have not read the above yet, but I *have* just modified/added some things to my previous document.

**Comment by gaow** [comment]
Thank you @stephens999 . I read your updates. I think it's nice and clear. We are mostly settled on terminology, if you can comment on the idea of `super-module` or `sub-pipeline`

```
super_module: module_1 * (module_2, module_3)
```
Or 
```
sub_pipeline: module_1 * (module_2, module_3)
```

I personally prefer `super-module`. Another big issue is that I need comments on the loss of succinctness  as a consequence of removing the block/step concept for cases such as the following (also see thread above). 

```
block_name:
  exec: 1.R, 2.R, 3.R, 4.R, 5.R
  params:
     K: 1, 2, 4, 50, 100, 200, 300
```

Seed issue and extracting info should go to separate tickets that I'm about to create.

**Comment by gaow** [comment]
Here is a elaborated example. With the current interface:

```
method_series_1:
  exec: 1.R, 2.R, 3.R, 4.R, 5.R
  params:
     K: 1, 2, 4, 50, 100, 200, 300
  return: something

method_series_2:
  exec: 1.R, 2.R, 3.R, 4.R, 5.R
  params:
     gamma: 0.1, 0.2, 0.4, 5, 10, 20, 30
  return: something

DSC:
   run: 
     pipeline_1: one_block * another_block * (method_series_1, method_series_2) * final_block 
     pipeline_2: ...
```

With proposed syntax that replaces the concept of block with modules and no multiple steps in the same module:

```
method_series_1:
  exec: 1.R
  params:
     K: 1
  return: something

method_series_2(method_series_1):
  exec: 2.R
  params:
    K: 1

(and you'll have 35 modules for K related parameters and another 35 modules for gamma related parameters).

DSC
   run:
    pipeline_1:   one_block * another_block * method_series_1 * final_block 
    pipeline_2: ...

(you'll have to write 70 pipelines)
```

**Comment by stephens999** [comment]
@gaow I'm not ready yet to talk about the YAML syntax. I want to get the terminology for 
*concepts* straight before getting to YAML syntax. I have thoughts on YAML syntax,
but all I want to say now is that i think they will likely require relatively small
changes to what you have now. I do not envisage the need for big changes.

Your last example "With proposed syntax ..." suggests you think I have proposed changes in YAML syntax  - I did not intend to propose any changes yet. Your current syntax looks largely great for
this example.  As you can see from my
more recent changes to the document, I'm still focussed on very basic concepts -
pipeline, module, parameters etc. 

Regarding multi-modules, you are right that this referred to the use of the same executable
with different parameters. I agree we want to keep the option of 
using a single name in this situation. But I'm not yet ready to fully explain all my thinking here, so
 I deliberately left this idea for further elaboration. For this reason I now removed this
terminology from my comment document. I will just say that I think
we can deal with this situation without an annotation step (leaving you to guess how :))

**Comment by gaow** [comment]
Thank you @stephens999 . Exactly I'm trying to guess how we can extract any situation we want without annotation step while keeping the DSC syntax concise. To me, multi-modules is the way to conciseness but I currently cannot think of an obvious way to extract info from mult-module without annotation. It's great you've got a way out!! I'll keep thinking while waiting on your updated post -- maybe I was too much constraint by having to worry about syntax and concepts at the same time.

The focus of this ticket is to get terminology straight, as you pointed out, and I totally agree. @pcarbo and I had a discuss earlier trying to sort this out. @pcarbo has used some formal literature to propose changes which he'll share after organizing the discussion notes.

**Comment by pcarbo** [comment]
I had a long discussion with Gao yesterday about this thread. We both agree that part of the difficulty here is that these discussions are not grounded in more precise (and often more familiar!) programming language terminology (e.g., "module" and "block" are not well-defined terms in this context, but a "function" is).

For me, most of my background on programming languages comes from [Structure and Interpretation of Computer Programs](https://mitpress.mit.edu/sites/default/files/6515.pdf) (the full text can be downloaded as a PDF if you click the link). Although this book is a bit dated, and many of the concepts are introduced more informally, it is a highly influential book, and I believe that the core concepts remain unchanged.

Here are my sugestions:

+ First of all, I can't see anywhere where we define what **is** a DSC. We know a DSC is a *computational process*. Can we define it more precisely than that? (A "computational experiment"?)

+ There was also a discussion about "block" vs. "module". From a programming languages point-of-view, "module" typically differs to a set of functions that would be packaged together, often within a common namespace (e.g., in Python, Java, C++). I agree that "module" seems natural because it connects to "modularity", but due to the historical usage of the word "module", I think it would add confusion. By contrast, "block" is an appropriate term because "code blocks" and "block structure" are often used as precursor concepts to "functions" and "abstraction". (Functions are almost always defined in "blocks" of code.) So the term "block" actually ties very closely to "function", and "modularity". I will use "block" for the rest of this post.

+ I also think it is useful to think about the DSC blocks as *functions*; that is, a block defines inputs and outputs, and a relationship between the two. Then we can think about the properties of a block in the DSC language, and how blocks are similar or different from functions in other languages. Consider:

    1. All variables either depend on the function parameters (more precisely, the binding of the function parameters to a given environment), or they don't. I think this might clarify the difference between what Gao calls "variables" (starting with $) and "block parameters", but Gao wasn't quite sure about this.

    2. All variables are *local* to the block; that is, the block inputs determine the dependencies between block variables (the exception is the set of global variables defined in the "DSC" block, but that can be treated as a block in which all blocks with mentioning global variables depend on it).

    3. One corollary of this is that all variables are evaluated with respect to a particular block and a particular environment. Therefore, we cannot ask, "what is the value of *x*" unless we specify the block and the environment (i.e., binding of block parameters) in which variable *x* is evaluated.

    4. Any valid DSC pipeline must have at least one block with no inputs (a "leaf").

    5. A block has *no side effects*; that is, the state is accessible only through the "return value".

    6. An important property of a block is that it has *limited compositionality*; that is, *a block cannot be defined in terms of other blocks*. This is in contrast to functions in most programming languages, in which we can define a function f3 that calls functions f1 and f2. Although this design decreases flexibility of the language, this was done deliberately to encourage good DSC design. The disadvantage is that there are some DSC patterns that cannot be expressed concisely without allowing for a language construct that defines a relationship between multiple functions. This is commonly called a "higher-order function". This has motivated the idea of having "super blocks" (or "super modules") that combine blocks in order to express the DSC more concisely.

    7. An important distinction between a DSC "block" and a "function" is that a block also has an extra property that determines the *domain*, or "possible values," for one or more variables that are defined locally within the block. This is important because the DSC *rules* (discussed below) define an evaluation rule that apply to all valid interpretations ("instantiations") of the variables. For example, supposing the rule `block1 * block2` is equivalent to evaluating all statements of the form `block2(y = block1(x = a),z = b)`, in which x and z are block parameters (or sets of parameters) that do not depend on any other blocks, y is a set bound variables, and a and b are valid instantiations of x and z, respectively. *This is a key feature of the DSC language that allows us to express computational experiments concisely.*

+ The "rule" statement defines a set of evaluation sequences for all valid instantiations of the block parameters. We have called each valid evaluation sequence a "pipeline". A pipeline also features: (1) an instantiation of all variables in all blocks included in the sequence; (2) a well-defined, local environment in which the variables in each block are evaluated.

Some of this may be a bit pedantic, but I think helps to clarify some of the language we use. For example, we cannot talk about instantiations of variables without associating the variables with (1) a block, and (2) an environment in which the variable is evaluated. 

To briefly reiterate our main aim, we would like to develop a language that can express common patterns in computational experiments ("DSCs") *concisely*, while *limiting flexibility* in order to encourage better DSC design.

**Comment by stephens999** [comment]
thanks @pcarbo
I agree with much of your comments here.  But I much prefer module to block
as a term.
I think it captures many of the features of our modules.

eg from Wikipedia

*Modular programming* is a software design
<https://en.wikipedia.org/wiki/Software_design> technique that emphasizes
separating the functionality of a program
<https://en.wikipedia.org/wiki/Computer_program> into independent,
interchangeable *modules*, such that each contains everything necessary to
execute only one aspect of the desired functionality.

**Comment by stephens999** [comment]
oh, and I'm deliberately holding off from defining a DSC for now!
I think we will get there!

**Comment by pcarbo** [comment]
Following up on my suggested programming language terminology, we have been using the following terms:

+ benchmark
+ task
+ (computational) step
+ class
+ super-module

Are any of these terms synonymous? How could we frame these terms using the terminology above?

**Comment by stephens999** [comment]
i think I'm going to try to avoid any of those terms for now.

I could see task or job might be useful to refer to an actual job running
on a computer system...

I think for me computational step has been retired in favor of module
instance or module.


Matthew

**Comment by pcarbo** [comment]
That works, too.

**Comment by gaow** [comment]
Discussion formalized at:

https://github.com/stephenslab/dsc-wiki/blob/master/development/finalized_terminology_and_extraction.md
